src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,366 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import os
+from typing import Any, Dict, List, Optional, Union
+
+import torch
+import torch.nn as nn
+from safetensors.torch import save_model  # type: ignore
+from transformers import PreTrainedModel
+
+from peft.tuners.lora.model import LoraModel
+from peft.utils.peft_types import PeftType
+
+from .. import lora
+from .classifier import InhibitorFlagPayload, Number, XLoraClassifier
+from .config import XLoraConfig
+from .layer import XLoRALinearLayer, XLoRAEmbeddingLayer, XLoRAConv2dLayer
+
+@staticmethod
+def apply_scalings_to_x(x: torch.Tensor, scalings_layer: torch.Tensor, adapter: int) -> torch.Tensor:
+    # scalings_layer = [batch_size, seq_len, n_classes]
+    scalings = scalings_layer[:, :, adapter].unsqueeze(-1)
+    # scalings_layer = [batch_size, seq_len, 1]
+    return x * scalings
+
+def convert_layers_to_xlora(
+    base: nn.Module,  # PeftModel
+    xloramodel: nn.Module, # XLoraModel
+    config: XLoraConfig,
+) -> int:
+    """
+    Returns the number of swapped layers.
+    """
+    total_swapped = 0
+
+    for module in base.modules():
+        if isinstance(module, lora.Linear):
+            new_layer = XLoRALinearLayer(
+                model=xloramodel,
+                target=module,
+                target_forward=module.forward,
+                layer_number=total_swapped,
+                config=config,
+            )
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+        elif isinstance(module, lora.Embedding):
+            new_layer = XLoRAEmbeddingLayer(
+                model=xloramodel,
+                target=module,
+                target_forward=module.forward,
+                layer_number=total_swapped,
+                config=config,
+            )
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+        elif isinstance(module, lora.Conv2d):
+            new_layer = XLoRAConv2dLayer(
+                model=xloramodel,
+                target=module,
+                target_forward=module.forward,
+                layer_number=total_swapped,
+                config=config,
+            )
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+            
+    return total_swapped
+
+
+class XLoraModel(LoraModel):
```

Comment:
Not sure if inheriting from `LoraModel`, vs `BaseTuner`, is really a good idea. Let's keep it as is for the moment but we may have to change this later.

Reply:
Yes, it is certainly not the most elegant approach.

Reply:
> About inheriting from LoraModel vs BaseTuner, it is certainly not the most elegant, but I think the only other way would be to encapsulate a LoraModel. This choice makes XLoraModel kind of complicated to understand, do you think it would be better to store a LoraModel?

Okay, then let's keep it as is for now.